### PR TITLE
Support flexible bound element types

### LINKED ISSUE #519: Generalize binding outside message channels

### PR DESCRIPTION
**Pull request by mbogoevici**
Fixes #519

Introduces some internal changes to the framework allowing the use
of other types than MessageChannel/SubscribableChannel as bindable
types (e.g. Flux, Observable, KStream, etc.)

- Modify BinderFactory to allow the retrieval and lookup of a
  binder not only by name, but also by binding target type
- Subsequent changes to ChannelBindingService and tests to account
  for the modified signature
- Introduce BindingTargetFactory as the contract for creating bound
  elements
- Remove any references to chanels and bound elements and use
  'binding target' systematically across the board

### CODE REVIEW

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BindingTargetFactory.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	 * Create an input binding target that will be bound via a corresponding binder
+	 * {@link org.springframework.cloud.stream.binder.Binder}.
+	 * @param name name of the binding target
+	 * @return binding target
+	 */
+	Object createInput(String name);
+
+	/**
+	 * Create an output binding target that will be bound via the message
+	 * channel {@link org.springframework.cloud.stream.binder.Binder}.
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BindingTargetFactory.java]: still refers to channel here

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java [hunk 2]
```diff
@@ -318,11 +326,11 @@ private Method checkProxy(Method methodArg, Object bean) {
...
 				throw new IllegalStateException(String.format(
-						"@StreamListener method '%s' found on bean target class '%s', " +
-								"but not found in any interface(s) for bean JDK proxy. Either " +
-								"pull the method up to an interface or switch to subclass (CGLIB) " +
-								"proxies by setting proxy-target-class/proxyTargetClass " +
-								"attribute to 'true'", method.getName(), method.getDeclaringClass().getSimpleName()), ex);
+						"@StreamListener method '%s' found on bean target class '%s', "
+								+ "but not found in any interface(s) for bean JDK proxy. Either "
+								+ "pull the method up to an interface or switch to subclass (CGLIB) "
+								+ "proxies by setting proxy-target-class/proxyTargetClass " + "attribute to 'true'",
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java]: unnecessary concatenation on the same line after some formatting changes

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/BindingServiceConfiguration.java [hunk 3]
```diff
@@ -171,17 +170,23 @@ public CompositeMessageConverterFactory compositeMessageConverterFactory() {
...
 	public static StreamListenerAnnotationBeanPostProcessor bindToAnnotationBeanPostProcessor(
 			@Lazy BinderAwareChannelResolver binderAwareChannelResolver,
 			@Lazy MessageHandlerMethodFactory messageHandlerMethodFactory) {
-		return new StreamListenerAnnotationBeanPostProcessor(binderAwareChannelResolver,
-				messageHandlerMethodFactory);
+		return new StreamListenerAnnotationBeanPostProcessor(binderAwareChannelResolver, messageHandlerMethodFactory);
+	}
+
+	@Bean
+	// provided for backwards compatibility scenarios w
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/BindingServiceConfiguration.java]: extraneous "w" or truncated comment?

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/internal/InternalPropertyNames.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package org.springframework.cloud.stream.internal;
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/internal/InternalPropertyNames.java]: missing copyright header

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/reflection/GenericsUtils.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
+package org.springframework.cloud.stream.reflection;
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/reflection/GenericsUtils.java]: missing copyright header

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/aggregate/AggregateApplicationUtils.java [hunk 6]
```diff
@@ -33,63 +34,52 @@
...
-
-	private static final String SPRING_CLOUD_STREAM_INTERNAL_PREFIX = "spring.cloud.stream.internal";
-
-	private static final String CHANNEL_NAMESPACE_PROPERTY_NAME =
-			SPRING_CLOUD_STREAM_INTERNAL_PREFIX + ".channelNamespace";
-
-	private static final String SELF_CONTAINED_APP_PROPERTY_NAME =
-			SPRING_CLOUD_STREAM_INTERNAL_PREFIX + ".selfContained";
-
 	public static final String INPUT_CHANNEL_NAME = "input";
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/aggregate/AggregateApplicationUtils.java]: should these be renamed... or always still *channels* in this case?

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/aggregate/AggregateApplicationUtils.java [hunk 7]
```diff
@@ -33,63 +34,52 @@
...
-				.registerShutdownHook(false)
-				.parent(parentContext);
+				.properties(
+						InternalPropertyNames.NAMESPACE_PROPERTY_NAME + "=" + namespace)
+				.registerShutdownHook(false).parent(parentContext);
 	}
 
-	static void prepareSharedChannelRegistry(SharedChannelRegistry sharedChannelRegistry,
+	static void prepareSharedBindingTargetRegistry(
+			SharedBindingTargetRegistry sharedChannelRegistry,
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/aggregate/AggregateApplicationUtils.java]: the method param name should probably change even if this impl is still currently opinionated toward channels

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/aggregate/SharedBindingTargetRegistry.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
+package org.springframework.cloud.stream.aggregate;
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/aggregate/SharedBindingTargetRegistry.java]: missing copyright header

**Comment by mbogoevici** [comment]
Addresses changes with another commit. I've reenabled our own checkstyle configuration which is more comprehensive than the one of Spring Cloud (and includes header validations) and more strict rules. About the latter, I'd like to re-enable all rules as it seems like some violations have accumulated over time, but we should do that as a separate PR.

File: spring-cloud-stream-reactive/src/test/java/org/springframework/cloud/stream/reactive/StreamListenerReactiveReturnWithFailureTests.java [hunk 9]
```diff
@@ -42,6 +40,8 @@
 import org.springframework.messaging.handler.annotation.SendTo;
 import org.springframework.messaging.support.MessageBuilder;
 
+import static org.assertj.core.api.Assertions.assertThat;
```
Comment (markfisher) [inline spring-cloud-stream-reactive/src/test/java/org/springframework/cloud/stream/reactive/StreamListenerReactiveReturnWithFailureTests.java]: I thought our guidelines were static imports first.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java [hunk 10]
```diff
@@ -37,22 +37,23 @@
  */
 public class BinderAwareChannelResolver extends BeanFactoryMessageChannelDestinationResolver {
 
-	private final ChannelBindingService channelBindingService;
+	private final BindingService bindingService;
 
-	private final BindableChannelFactory bindableChannelFactory;
+	private final AbstractBindingTargetFactory<? extends MessageChannel> boundElementFactory;
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java]: "boundElementFactory" should be replaced with "bindingTargetFactory" throughout this class

File: spring-cloud-stream-reactive/src/test/java/org/springframework/cloud/stream/reactive/StreamListenerReactiveReturnWithFailureTests.java (hunk 9 above)
Comment (mbogoevici) [inline spring-cloud-stream-reactive/src/test/java/org/springframework/cloud/stream/reactive/StreamListenerReactiveReturnWithFailureTests.java]: The Spring Framework Code Style guidelines has static imports at the end: https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-Code-Style#import-statements

The intent is ban all static imports bar from JUnit and AssertJ assertions.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java [hunk 11]
```diff
@@ -254,31 +262,29 @@ else if (arguments.length == 1 && StringUtils.hasText(inboundName)) {
...
 			Assert.isTrue(StringUtils.isEmpty(defaultOutputChannel),
-					"An output channel cannot be specified for a method that " +
-							"does not return a value");
+					"An output channel cannot be specified for a method that " + "does not return a value");
 		}
 		else {
 			Assert.isTrue(!StringUtils.isEmpty(defaultOutputChannel),
-					"An output channel must be specified for a method that " +
-							"can return a value");
+					"An output channel must be specified for a method that " + "can return a value");
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java]: no longer need concatenation now that it's one line

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/BindingServiceConfiguration.java [hunk 12]
```diff
@@ -140,17 +139,17 @@ public ContextStartAfterRefreshListener contextStartAfterRefreshListener() {
 	}
 
 	@Bean
-	public BinderAwareChannelResolver binderAwareChannelResolver(ChannelBindingService channelBindingService,
-			BindableChannelFactory bindableChannelFactory, DynamicDestinationsBindable dynamicDestinationsBindable) {
-		return new BinderAwareChannelResolver(channelBindingService, bindableChannelFactory,
-				dynamicDestinationsBindable);
+	public BinderAwareChannelResolver binderAwareChannelResolver(BindingService bindingService,
+			AbstractBindingTargetFactory<? extends MessageChannel> boundElementFactory,
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/BindingServiceConfiguration.java]: param name should be `bindingTargetFactory`

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/BindingServiceProperties.java [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+		for (Map.Entry<String, BindingProperties> entry : this.bindings.entrySet()) {
+			properties.put(entry.getKey(), entry.getValue().toString());
+		}
+		for (Map.Entry<String, BinderProperties> entry : this.binders.entrySet()) {
+			properties.put(entry.getKey(), entry.getValue());
+		}
+		return properties;
+	}
+
+	public ConsumerProperties getConsumerProperties(String inputChannelName) {
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/BindingServiceProperties.java]: should avoid "channel" here (and next line)?

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/BindingServiceProperties.java [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+		if (consumerProperties.getInstanceCount() < 0) {
+			consumerProperties.setInstanceCount(this.instanceCount);
+		}
+		if (consumerProperties.getInstanceIndex() < 0) {
+			consumerProperties.setInstanceIndex(this.instanceIndex);
+		}
+		return consumerProperties;
+	}
+
+	public ProducerProperties getProducerProperties(String outputChannelName) {
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/BindingServiceProperties.java]: should avoid "channel" here (and next line)?

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/BindingServiceProperties.java [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+		Assert.notNull(outputChannelName, "The output channel name cannot be null");
+		ProducerProperties producerProperties = getBindingProperties(outputChannelName)
+				.getProducer();
+		if (producerProperties == null) {
+			producerProperties = new ProducerProperties();
+		}
+		return producerProperties;
+	}
+
+	public BindingProperties getBindingProperties(String channelName) {
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/BindingServiceProperties.java]: "channelName" here and other methods below

**Comment by markfisher** [comment]
LGTM; rebased, squashed, and merged as https://github.com/spring-cloud/spring-cloud-stream/commit/8ac71c7b57bc00b291e84ec345235be4ac57ba32

**Pull request closed without merging**